lib/verification.rs

```diff
@@ -0,0 +1,209 @@
+/// public key, and code archive.
+///
+/// # Arguments
+///
+/// * `deploy_hash` - The hash of the deployed contract.
+/// * `public_key` - The public key associated with the contract.
+/// * `base_url` - The base path of the verification URL.
+/// * `verbosity` - The verbosity level of the verification process.
+///
+/// # Returns
+///
+/// The verification details of the contract.
+pub async fn send_verification_request(
+    key: Key,
+    base_url: &str,
+    code_archive: String,
+    verbosity: Verbosity,
+) -> Result<VerificationDetails, Error> {
+    let verification_request = VerificationRequest {
+        deploy_hash: key,
+        code_archive,
+    };
+
+    let mut headers = HeaderMap::new();
+    headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
+
+    let Ok(http_client) = ClientBuilder::new()
+        .default_headers(headers)
+        // https://github.com/hyperium/hyper/issues/2136
+        .pool_max_idle_per_host(0)
+        .user_agent("casper-client-rs")
+        .build()
+    else {
+        eprintln!("Failed to build HTTP client");
+        return Err(Error::FailedToConstructHttpClient);
+    };
+
+    if verbosity == Verbosity::Medium || verbosity == Verbosity::High {
+        println!("Sending verification request");
+    }
+
+    let url = base_url.to_string() + "/verification";
+    let response = match http_client
+        .post(url)
+        .json(&verification_request)
+        .send()
+        .await
+    {
+        Ok(response) => response,
+        Err(error) => {
+            eprintln!("Cannot send verification request: {error:?}");
+            return Err(Error::ContractVerificationFailed);
+        }
+    };
+
+    match response.status() {
+        StatusCode::OK => {
+            if verbosity == Verbosity::Medium || verbosity == Verbosity::High {
+                println!("Sent verification request",);
+            }
+        }
+        status => {
+            eprintln!("Verification failed with status {status}");
+        }
+    }
+
+    wait_for_verification_finished(base_url, &http_client, key, verbosity).await;
+
+    if verbosity == Verbosity::Medium || verbosity == Verbosity::High {
+        println!("Getting verification details...");
+    }
+
+    let url = base_url.to_string() + "/verification/" + &key.to_formatted_string() + "/details";
+    match http_client.get(url).send().await {
+        Ok(response) => response.json().await.map_err(|err| {
+            eprintln!("Failed to parse JSON {err}");
+            Error::ContractVerificationFailed
+        }),
+        Err(error) => {
+            eprintln!("Cannot get verification details: {error:?}");
+            Err(Error::ContractVerificationFailed)
+        }
+    }
+}
+
+/// Waits for the verification process to finish.
+async fn wait_for_verification_finished(
+    base_url: &str,
+    http_client: &Client,
+    key: Key,
+    verbosity: Verbosity,
+) {
+    let mut retries = MAX_RETRIES;
```

Comment:
Are there tests that cover this logic of retrying 1200 times every 3 seconds ? is there a way to limit this load ?

10 attempts seems more reasonable with time extending between attempts

const MAX_RETRIES: u32 = 10;
const BASE_DELAY: u64 = 3;
const MAX_DELAY: u64 = 300;

```delay = std::cmp::min(delay * 2, MAX_DELAY); ```

Would this be possible to unit test this function ?